### DISCUSSION TITLE
Guide not working for me

### DISCUSSION THREAD

**Discussion by danielniccoli** [diesel-rs/diesel#4089] [Q&A]
I'm at this step:

```
use schema::users::dsl::*;
insert_into(users).default_values().execute(conn)
```

https://diesel.rs/guides/all-about-inserts.html

It doesn't work for me: `cannot find function `insert_into` in this scope`

![grafik](https://github.com/diesel-rs/diesel/assets/2971735/f31e2a30-cc1d-4358-9d5e-0ddc0a7e5522)

How to fix?

**Comment by weiznich** [comment]
The error message already points out the correct solution: 

> consider importing this function: `use diesel:: insert_ into;`

More general speaking: The underlying code for almost all of the guides is tested in our CI so it's guaranteed to work. In this case the relevant complete code is linked from the end of the guide for the different backends.

**Reply by danielniccoli** [reply]
> The error message already points out the correct solution:

Which made me question if the guide is wrong, or was not updated. So here I am.

>  it's guaranteed to work

It did not work, though, did it?

>  the relevant complete code is linked from the end of the guide for the different backends.

Why write a guide, that contains code that will throw an error, when at its end you refer to the sample project anyway? That defeats the entire purpose of having a guide in the first place.

**Reply by weiznich** [reply]
It is common for code snippets in guides to skip lines that are not directly relevant to what is explained in that specific context. This is done to keep the code parts short. We do explicitly link the example projects for providing the full context for everything. They are there to be used. 

Imports are one of those things that are commonly skipped there, as they are such a fundamental part of the language that it is obvious for any user with basic rust knowledge that they need to import something to actually use this function. As this guide requires at least basic understanding of rust anyway I personally see no problem with not having that important in each code block. 

That all written: I personally do not plan to change the guide just because one person complained about this, as the guide exists for quite a few years now and it seems to work for the majority of our users and because the relevant error message is really clear about what's wrong. I would nevertheless accept a PR that changes the first code example in the guide.

**Reply by ksaadDE** [reply]
> I personally do not plan to change the guide just because one person complained about this

Yet again you are ignoring that people like me and many others have complained about that. What about questioning your own approach towards us as community?

>  fundamental part of the language that it is obvious for any user with basic rust knowledge

That ad hominem is unbelievable. Yet again you are insulting your whole user community.


> I personally see no problem 

Oh sure anecdotal evidence absolutely proves your point and wishes away all the sound critique. We are not checking whether our assumptions are right, instead we are being you just saying "yeah no I have no problem, so you don't have a problem either"

> I would nevertheless accept a PR that changes the first code example in the guide.

I thought we all are too stupid and not having basic RUST knowledge. So how should we fix YOUR guide? Your logic is insane.

Sorry but if you really want to establish some professionalism I would be really happy to see it! If you can not due to autism or something like that, try to accept or ask for help from others. 🙏

**Reply by ksaadDE** [reply]
BTW I went on the guide, pressed CRTL+F and entered the snippet (`use diesel:: insert_ into;`) and variants of it, just to be sure. Guess what? Your proposed "it's already in the guide"-solution is not even in there. Stop being rude. Just make the guide more clearer by helping people to solve the issue and then clarify it in the guide. Use the people asking such questions to improve your guide. Really, it's not that hard.

**Reply by weiznich** [reply]
I just want to make a few things very clear here:

1. This behaviour is not welcome here. If you continue to insult people I will block you from the repo. Consider this as your last warning. 
2. This is free software maintained in my free time. I own you nor anybody else nothing. The license explicitly states that "THE SOFTWARE IS PROVIDED AS IS". You are free to use it but you are not entitled that any of the maintainers does something for you, not even answering any of you questions. 

Given these points and my own different priorities: I won't spend time on fixing these minor issues in the guide. As written before: If it bothers you feel free to submit a PR.

**Reply by ksaadDE** [reply]
@weiznich  So all in all you are wondering why people suck up your time and complaining here, if you are putting up a partially broken and mostly hard to understand guide? 

If you do not want insults, better stop insulting the community, it's that simple. 

Side note: We just put up sound critique to about your behave and what could be improved. It rather appears you have a hard time to stand that fact. So that's probably what you perceive as "insult". But in reality nobody insulted you.

Your excuse "provided as is" does not explain why you have those significant issues more than other projects that we all came across. Other people are happy to clarify and to fix those literally basic things. 

I mean you do not have to provide the project to us as FOSS community. Then someone else replaces you and your project. ^^

BTW: the same way it's "provided as is", we do not have to contribute anything, no PRs, no feedback, no bug reports for you. Have fun doing it all on your own... ;)

**Reply by ksaadDE** [reply]
Moved my reply to: https://github.com/diesel-rs/diesel/discussions/4099, I was pointing out how many times the guide is unclear or lacks information issue came up. I went through the past Q&A's by the keyword "guide" and summarized each, were I found that issue.

**Reply by ksaadDE** [reply]
> obvious for any user with basic rust knowledge

I am still disagreeing to that notion. It insults the user, especially beginners.

>   I personally see no problem with not having that important in each code block.

While it may look cleaner for advanced users, knowing what to import is crucial for beginners. They are not used to the customs. Most people learn by doing, so they spin up a quick playground project, getting used to the customs by trying to go hands-on.

> I personally do not plan to change the guide just because one person complained about this [...]  guide exists for [...]  years now and [... seems to work...] for the majority [...]

Multiple people complained, which is summarized in #4099 

> Why write a guide, that contains code that will throw an error, when at its end you refer to the sample project anyway? That defeats the entire purpose of having a guide in the first place.

+1
